app/src/hog.c

```diff
@@ -156,28 +157,120 @@ struct bt_conn *destination_connection() {
     return conn;
 }
 
+#define HOG_STACK_SIZE 512
+#define HOG_PRIORITY 5
+
+K_THREAD_STACK_DEFINE(hog_q_stack, HOG_STACK_SIZE);
+
+struct k_work_q hog_work_q;
+
+#define HOG_REPORT_QUEUE_SIZE 5
+
+K_MSGQ_DEFINE(zmk_hog_keyboard_msgq, sizeof(struct zmk_hid_keyboard_report_body),
+              HOG_REPORT_QUEUE_SIZE, 4);
+
+void send_keyboard_report_callback(struct k_work *work) {
+    struct zmk_hid_keyboard_report_body report;
+
+    while (k_msgq_get(&zmk_hog_keyboard_msgq, &report, K_NO_WAIT) == 0) {
+        int err;
+        struct bt_conn *conn = destination_connection();
+        if (conn == NULL) {
+            return;
+        }
+
+        struct bt_gatt_notify_params notify_params = {
+            .attr = &hog_svc.attrs[5],
+            .data = &report,
+            .len = sizeof(struct zmk_hid_keyboard_report_body),
+        };
+
+        err = bt_gatt_notify_cb(conn, &notify_params);
+        if (err) {
+            LOG_ERR("Error notifying %d", err);
+        }
+
+        bt_conn_unref(conn);
+    }
+}
+
+K_WORK_DEFINE(hog_keyboard_work, send_keyboard_report_callback);
+
 int zmk_hog_send_keyboard_report(struct zmk_hid_keyboard_report_body *report) {
-    struct bt_conn *conn = destination_connection();
-    if (conn == NULL) {
-        return -ENOTCONN;
+    struct zmk_hid_keyboard_report_body discarded_report;
+
+    int err = k_msgq_put(&zmk_hog_keyboard_msgq, report, K_MSEC(100));
+    if (err) {
+        switch (err) {
+        case -EAGAIN:
+            LOG_WRN("keyboard message queue full, popping first message and queueing again");
+            k_msgq_get(&zmk_hog_keyboard_msgq, &discarded_report, K_NO_WAIT);
+            return zmk_hog_send_keyboard_report(report);
+        default:
+            LOG_WRN("Failed to queue keyboard report to send (%d)", err);
+            return err;
+        }
     }
 
-    LOG_DBG("Sending to NULL? %s", conn == NULL ? "yes" : "no");
+    k_work_submit_to_queue(&hog_work_q, &hog_keyboard_work);
 
-    int err = bt_gatt_notify(conn, &hog_svc.attrs[5], report,
-                             sizeof(struct zmk_hid_keyboard_report_body));
-    bt_conn_unref(conn);
-    return err;
+    return 0;
 };
 
+K_MSGQ_DEFINE(zmk_hog_consumer_msgq, sizeof(struct zmk_hid_consumer_report_body),
```

Comment:
Yeah, I pondered just queueing the notify params, but those take a *pointer* to the notify data, so doing that won't snapshot the hid report structs like this implementation does. So we need a queue for each one, since queue's need to define a size for each entry.

That or some union crazy. I'd favor leaving as is for now, and possibly refactor later.